contracts/Everlong.sol

```diff
@@ -307,35 +299,99 @@ contract Everlong is IEverlong {
-    function rebalance() public override {
+    function rebalance() external onlyAdmin {
+        _rebalance(
+            IEverlong.RebalanceOptions({
+                spendingOverride: 0,
+                minOutput: 0,
+                minVaultSharePrice: 0,
+                positionClosureLimit: 0,
+                extraData: ""
+            })
+        );
+    }
+
+    // NOTE: Errors from hyperdrive are not handled. The keeper must configure
+    //       the correct parameters to avoid issues with insufficient liquidity
+    //       and running out of gas from mature position closures.
+    //
+    /// @notice Rebalance the everlong portfolio by closing mature positions
+    ///         and using the proceeds over target idle to open new positions.
+    /// @param _options Options to control the rebalance behavior.
+    function rebalance(
+        IEverlong.RebalanceOptions memory _options
+    ) external onlyAdmin {
+        _rebalance(_options);
+    }
+
+    // NOTE: Errors from hyperdrive are not handled. The keeper must configure
+    //       the correct parameters to avoid issues with insufficient liquidity
+    //       and running out of gas from mature position closures.
+    //
+    /// @dev Rebalance the everlong portfolio by closing mature positions
+    ///         and using the proceeds over target idle to open new positions.
+    /// @param _options Options to control the rebalance behavior.
+    function _rebalance(IEverlong.RebalanceOptions memory _options) internal {
         // Early return if no rebalancing is needed.
         if (!canRebalance()) {
             return;
         }
 
-        // Calculate the new portfolio value and save it.
-        _totalAssets = _calculateTotalAssets();
-
         // Close matured positions.
-        _closeMaturedPositions();
+        _closeMaturedPositions(_options.positionClosureLimit);
+
+        // If Everlong has sufficient idle, open a new position.
+        if (canOpenPosition()) {
+            // Calculate how much idle to spend on the position.
+            uint256 balance = IERC20(_asset).balanceOf(address(this));
+            uint256 target = targetIdleLiquidity();
+            uint256 toSpend = balance - target;
+            if (_options.spendingOverride != 0) {
+                // Spending override must not be less than Hyperdrive's
+                // minimum transaction amount.
+                if (
+                    _options.spendingOverride <
+                    IHyperdrive(hyperdrive)
+                        .getPoolConfig()
+                        .minimumTransactionAmount
+                ) {
+                    revert SpendingOverrideTooLow();
+                }
+
+                // Spending override must not be greater than the difference
+                // between Everlong's balance and the max idle liquidity.
+                if (_options.spendingOverride > toSpend) {
+                    revert SpendingOverrideTooHigh();
+                }
```

Comment:
So this is the case where the spending limit is higher than the amount of idle to spend. Why do we need to revert here? This seems like it could result in DOS issues.

If we think of this as a `spendingLimit` rather than `spendingOverride` (which seems more consistent with the field's documentation), we could just use the smaller of `toSpend` and `_options.spendingOverride` (or `_options.spendingLimit`).

Reply:
Love this, going to keep it so that 0 means unlimited